go-controller/pkg/ovn/policy_common.go

```diff
@@ -165,6 +168,201 @@ func (gp *gressPolicy) delAddressSet(hashedAddressSet string) (string, string, b
 	return oldL3Match, gp.getL3MatchFromAddressSet(), true
 }
 
+func (oc *Controller) handlePeerPodSelectorAddUpdate(np *namespacePolicy,
+	addressMap map[string]bool, addressSet string, obj interface{}) {
+
+	pod := obj.(*kapi.Pod)
+	podAnnotation, err := util.UnmarshalPodAnnotation(pod.Annotations["ovn"])
+	if err != nil {
+		return
+	}
+	ipAddress := podAnnotation.IP.IP.String()
+	if addressMap[ipAddress] {
+		return
+	}
+
+	np.Lock()
+	defer np.Unlock()
+	if np.deleted {
+		return
+	}
+
+	addressMap[ipAddress] = true
+	addresses := make([]string, 0, len(addressMap))
+	for k := range addressMap {
+		addresses = append(addresses, k)
+	}
+	oc.setAddressSet(addressSet, addresses)
+
+}
+
+func (oc *Controller) handlePeerPodSelectorDelete(np *namespacePolicy,
+	addressMap map[string]bool, addressSet string, obj interface{}) {
+
+	pod := obj.(*kapi.Pod)
+	podAnnotation, err := util.UnmarshalPodAnnotation(pod.Annotations["ovn"])
+	if err != nil {
+		return
+	}
+	ipAddress := podAnnotation.IP.IP.String()
+
+	np.Lock()
+	defer np.Unlock()
+	if np.deleted {
+		return
+	}
+
+	if !addressMap[ipAddress] {
+		return
+	}
+
+	delete(addressMap, ipAddress)
+
+	addresses := make([]string, 0, len(addressMap))
+	for k := range addressMap {
+		addresses = append(addresses, k)
+	}
+	oc.setAddressSet(addressSet, addresses)
+}
+
+func (oc *Controller) handlePeerPodSelector(
+	policy *knet.NetworkPolicy, podSelector *metav1.LabelSelector,
+	addressSet string, addressMap map[string]bool, np *namespacePolicy) {
+
+	h, err := oc.watchFactory.AddFilteredPodHandler(policy.Namespace,
+		podSelector,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				oc.handlePeerPodSelectorAddUpdate(np, addressMap, addressSet, obj)
+			},
+			DeleteFunc: func(obj interface{}) {
+				oc.handlePeerPodSelectorDelete(np, addressMap, addressSet, obj)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				oc.handlePeerPodSelectorAddUpdate(np, addressMap, addressSet, newObj)
+			},
+		}, nil)
+	if err != nil {
+		logrus.Errorf("error watching peer pods for policy %s in namespace %s: %v",
+			policy.Name, policy.Namespace, err)
+		return
+	}
+
+	np.podHandlerList = append(np.podHandlerList, h)
+}
+
+func (oc *Controller) handlePeerNamespaceAndPodSelector(
+	policy *knet.NetworkPolicy,
+	namespaceSelector *metav1.LabelSelector,
+	podSelector *metav1.LabelSelector,
+	addressSet string,
+	addressMap map[string]bool,
+	np *namespacePolicy) {
+
+	namespaceHandler, err := oc.watchFactory.AddFilteredNamespaceHandler("",
+		namespaceSelector,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				namespace := obj.(*kapi.Namespace)
+				np.Lock()
+				alreadyDeleted := np.deleted
```

Comment:
Theoretically, you should be comparing the UIDs to ensure this isn't a new namespace. We can let this slide for now (I know this is a bigger refactor)

Reply:
@squeed also no change from previous, it's just moving code.